### ISSUE TITLE
RDS update stuck on boot

### ISSUE THREAD

**Issue by snowp** [envoyproxy/envoy#3399]
Using envoy at 71c0fd695e992e21d2e0439eb14b449fa0b8aea4 and java-control-plane we've run into an issue where envoy fails to load all the RDS resources at boot. 

Envoy is configured with two listeners, using two route tables (`standard_egress`/`standard_ingress`). Envoy begins by requesting the routes for *just* `standard_egress`, and gets this back with version = X. Following this it requests `[standard_egress, standard_ingress]` but since the nonce is the same the control plane ignores the request. Finally, envoy requests `[standard_egress, standard_ingress]` again with version X. Because the version matches what the control plane has, it doesn't send the requested resources right away, instead waiting for the next time the resources are updated. This leaves Envoy without a route table for `standard_ingress`, resulting in 404s.

Raw logs from the control plane: 
```
2018-05-16T08:10:52,156  INFO [rpc-server-dispatcher-9] server.DiscoveryServer - [1] request type.googleapis.com/envoy.api.v2.RouteConfiguration[standard_egress] with nonce  from version
2018-05-16T08:10:52,158  INFO [rpc-server-dispatcher-9] cache.SimpleCache - responding for type.googleapis.com/envoy.api.v2.RouteConfiguration from node slug--LOCAL at version  with version 0f9e47ac-faba-4b9f-a3bd-5e94c6b7dc35
2018-05-16T08:10:52,158  INFO [rpc-server-dispatcher-9] server.DiscoveryServer - [1] response type.googleapis.com/envoy.api.v2.RouteConfiguration with nonce 1 version 0f9e47ac-faba-4b9f-a3bd-5e94c6b7dc35
2018-05-16T08:10:52,162  INFO [rpc-server-dispatcher-10] server.DiscoveryServer - [1] request type.googleapis.com/envoy.api.v2.RouteConfiguration[standard_ingress, standard_egress] with nonce  from version
2018-05-16T08:10:52,168  INFO [rpc-server-dispatcher-12] server.DiscoveryServer - [1] request type.googleapis.com/envoy.api.v2.RouteConfiguration[standard_ingress, standard_egress] with nonce 1 from version 0f9e47ac-faba-4b9f-a3bd-5e94c6b7dc35
2018-05-16T08:10:52,170  INFO [rpc-server-dispatcher-12] cache.SimpleCache - open watch 5 for type.googleapis.com/envoy.api.v2.RouteConfiguration[standard_ingress, standard_egress] from node slug--LOCAL for version 0f9e47ac-faba-4b9f-a3bd-5e94c6b7dc35
```

What is the expected way of handling this? Not clear to me whether this is a control plane bug or an envoy bug - should the control plane update its version when new resources are requested? Or should envoy reset its version when it sends a request with new resources? or maybe wait until it knows what resources to request before sending the xDS request?

**Comment by mattklein123** [comment]
@snowp sorry to clarify you are using ADS, right?

**Comment by snowp** [comment]
Yes, this is using ADS.

**Comment by htuch** [comment]
Envoy should be pausing RDS updates during LDS config updates such that it presents only the `[standard_egress, standard_ingress]` request. I'd be interested in knowing why we're seeing multiple RDS requests initially.

**Comment by htuch** [comment]
Yeah, please do this, I agree that this is the correct behavior.

**Comment by htuch** [comment]
@mpuncel not sure I follow that example. In the case of an LDS update, we halt any new RDS requests until the full LDS update has been processed, see https://github.com/envoyproxy/envoy/blob/c7ba07ea3a0a49fec1ac24c763eb9f0a2e13c06b/source/server/lds_api.cc#L45. Only after the full LDS update is processed will RDS do something new, once we have the new route resource hint.

@snowp BTW, we probably have the same issue for static clusters and EDS.

**Comment by mpuncel** [comment]
lets say there's a listener named L1 that has a route config named R1. During steady state, I'd expect there to be an LDS discovery request (no resource hints) and an RDS discovery request with a resource hint for R1. If L2 is then added with route config R2 via the control plane, my understanding is that the current implementation of java-control-plane would send a discovery response for LDS and RDS simultaneously, which creates a race between RDS sending another request with just the R1 resource hint (perhaps before lds_api.cc has even seen the LDS response) or the LDS response being processed and the next RDS requests having R1 and R2 as hints.

I think I might have answered my own question though, I think the control plane should be able to notice that the R2 resource hint was added and send another response even before version change. I'm not sure java-control-plane does this right now, because it seems like that would have been another resolution to the problem snowp saw.

Basically i'd expect the `2018-05-16T08:10:52,168  INFO [rpc-server-dispatcher-12] server.DiscoveryServer - [1] request type.googleapis.com/envoy.api.v2.RouteConfiguration[standard_ingress, standard_egress] with nonce 1 from version 0f9e47ac-faba-4b9f-a3bd-5e94c6b7dc35` request to get a response because standard_ingress hadn't been requested before. So maybe there's a bug in java-control-plane as well?
